### PR TITLE
Dungeons 2.0 Complete!

### PR DESCRIPTION
**Pull request by NoiSek**
We've come a long way, boys, but Dungeons 2.0 is finally complete.

Dungeons is by no means complete even for the base framework that it is intended to be, but this milestone marks the point at which we have a viable base and the tools necessary to start building out more complex features as necessary.

Dungeons 1.x:

Dungeons 2.x:

**More unlisted file changes located here:** https://github.com/LikeMyBread/Saylua/pull/44

### Architectural Changes

Dungeons now utilizes Pixi.js, which speaks for itself for the most part. See the code to learn more.

In addition, we are now using Redux for the majority of our state management to handle the rising complexity of game interactions somewhat more sanely. As part of that, we get history traversal and deep inspection abilities with the Redux Devtools for free. [Chrome extension is here](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd).

*Notes on using the extension*:
Recording tree state to history with Redux can get quite slow due to the number of objects our game state has, if you begin to experience extremely slow movements 'pause recording' or close the inspector.

### Entity / Tile / Item (?) Schemas

Entities and Tiles are now defined with a more reusable schema, and are auto-generated from files -> DB using the default provisioner.

Example enemy schema:

`schema/entity/enemy_slime.json`
```javascript
{
  "name":        "Primordial Ooze",
  "description": "Sluggish beastie. Nasty things.",
  "type":        "enemy",
  "events":       {
      "$onIdle": "script_default_enemy_idle"
  },
  "meta":        {
      "maxHP": 10
  }
}
```

`schema/script/default_enemy_idle.j`
```javascript
console.log("Hello!");
```

Currently, the schema only supports using dynamic variables ($) for the names of matching scripts. In the future this may support more complex interactions between entities such as inheritance.

### Scripting Engine

The heart of the Dungeons 2.0 milestone, every other feature is made possible by this.

This PR introduces a scripting system that takes text blobs provided by the database and ties them to entities and tiles for execution at runtime. As an example, the hardcoded event 'onIdle' is called for every tile and entity within the current (view radius * 3) and allows them to specify a script to use when idling while the player is nearby.

To facilitate the use of these scripts, the following special variables are currently implemented and will be analyzed at compile time to determine the proper values to provide:

- `$this` => A reference to the current actor
- `$tileLayer` => A reference to a live copy of the tile array
- `$entityLayer` => A reference to a live copy of the entity array 
- `$player` => A reference to the player character
- `$tiles` => A reference to the tile set
- `$entities` => A reference to the entity set
- `$entity_nearest` => 'Magic' variable that returns the closest nearby entity
- `$entity_nearest_@type` => 'Magic' variable that returns the closest nearby entity of a given type
- `$location` => A dict of the current {x, y} grid location
- `__distance` => Helper method to measure the distance between two points
- `__isObstacle` => Helper method to determine whether a given (x, y) coordinate is an obstacle tile (or at least is telling us that it is)
- `__log` => Helper method to write to the event log. Currently just a wrapper around console.log.
- `__moveTo` => Returns a list of coordinates necessary to reach point B from point A, calculated with A*
- `__rand` => RNG range method
- `__debugMove` => Temporary move method
- `__debugAttack` => Temporary attack method

For the tl;dr of how this process works:

- Engine determines if the target, or the parent target possesses an event that matches the event it is currently trying to trigger
- Engine retrieves the text blob of the script, analyzes it's contents, and determines a list of necessary requirements
- Engine evalutes the textblob as a stored function, and caches a reference to the compiled script along with a list of the necessary requirements
- Engine reads the requirements, injects live copies of game state objects as necessary, and executes the script

More can be said on this, but try it for yourself.

### Debug Tools

Probably exactly what you think. Adds a few useful tools to help with dungeons development.

**General Debug Functions**

Easy to extend, allows the developer to add toggleable checks and useful shortcuts in a fairly straightforward fashion.

**Script Editor**

Allows the developer to test out the scripting capabilities of various things in real time, rather than having to rewrite them from the schema or DB.

Currently only supports the editing of pre-existing scripts, but will allow developers to add new scripts to arbitrary things in a future update.

Note: Does NOT save these changes. Liberal copy / pasting is important.

**Entity / Tile Summoner**

Loads the full list of entities / tiles from the server on the fly, allowing you to place them within the current Dungeon and test in-place without refreshing. Paired with the script editor this will become very useful in the future.

### CODE REVIEW

File: saylua/modules/explore/dungeons/maps/meta/__init__.py [hunk 1]
```diff
@@ -1,41 +1,6 @@
...
-        },
-        'meta':        { 'maxHP': 10 }
-    },
-    {
-        'id':          '0x3000',
-        'description': 'Ah, the classic three-timber dungeon door. Generic, but quite homely. You can\'t deny that it has an almost...adoorable look to it.',
```
Comment (LikeMyBread) [inline saylua/modules/explore/dungeons/maps/meta/__init__.py]: This is just too much, the pun level is insane.

File: saylua/modules/explore/dungeons/schema/entity/default_door.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+    "name":        "Wooden Door",
+    "description": "Ah, the classic three-timber dungeon door. Generic, but quite homely. You can\"t deny that it has an almost...adoorable look to it.",
```
Comment (LikeMyBread) [inline saylua/modules/explore/dungeons/schema/entity/default_door.json]: This is just too much, the pun level is insane.

**Comment by tiffz** [comment]
Man, this is extremely cool. One initial thought before I dive in further to this. 

I admit the scripts stored in the database sort of scare me because they seem like an XSS liability and it does sort of bother me that it would potentially be a lot of code that ends up being left non version controlled (not unit tested?), etc. 

I remember I thought of doing something similar for items (Python code that gets executed in specific item use cases) and entities inside of villages (JavaScript that gets executed on things like click, hover, etc). My thought back then was essentially having a folder of scripts in the source for these things and referencing the filename of the script or something in the database. 

Thoughts?

**Comment by NoiSek** [comment]
I've got a ticket open (#50) to address the non-version controlled thing, I think it's very important to have the relationship between the DB and the schema be bi-directional rather than one-way for many reasons-- just haven't put in the effort to make it a reality yet mostly as I didn't think it was immediately important.

As far as the XSS thing, it's a potential window for it from our end yeah, but only if we decide to XSS users ourselves or greenlight a PR that does. From the user end I don't really consider it more of a risk, as everything the scripting engine does is already 100% accomplishable by just opening the dev console really. I don't think there is an avenue by which a potential exploitee would affect any user other than themselves.

As long as we're serving over HTTPS and users can trust the source serving them files, I think we're fine on that front, but that goes for anything really.

**Comment by tiffz** [comment]
That's understandable. What are the advantages of making the DB and the schema has a bi-directional relationship? 

I am a bit curious because I did go through this thought process a little bit back when I thinking about how to extend items to have unique functionality and couldn't figure out a way to do this that didn't feel slightly terrible haha. Referencing scripts in the database will also be important for us to do for features like the quest engine. For some reason I feel like I would be a bit more hesitant about storing code which did things like grant users rewards or otherwise edited the database in this sort of format, but I'm not sure if this is a rational concern. 

Yeah, XSS is only a vulnerability really I guess if someone got access to our database directly or if inputs into the dungeon script were exposed somehow. Minor risk but not a dealbreaker.

**Comment by NoiSek** [comment]
The advantage of having schema work in two directions rather than one is mostly in being able to add / edit them from the admin interface and then write them back to files.  Easier to make a set of changes that way than to wipe the DB and re-provision every time you modify or create item data. For that to be a thing, however, we also have to make sure that all instances of Dungeons operate from the same data rather than becoming progressively more out of sync with their db-only models (and that's where the DB -> File serialization comes in).

The script engine is also not really meant to handle those sorts of external relations (rewards, DB manipulation, handling Saylua data), that part would be managed by the core client code. The scripting engine is only really meant to govern the immediate in-dungeon interactions between game entities.

Were you to work on the unique RNG item drop feature, for example, you wouldn't script the entity to give the user an item when it died; you would instead have the entity emit an "I'm dead boys" signal and the client would ingest it to handle item drops.

In a theoretical implementation, the core code would see that signal, log the event, POST the entity's ID to an endpoint that chooses from a random list, and then return said item.

With the addition of a token to prevent spoofing, the roundtrip on that request would probably be ~80ms so we'd have to spawn a chest immediately and then put something in the chest afterwards if we wanted to make the delay less noticeable. 

Note: that theoretical use-case is only for items that you get to keep outside of Dungeons, common items that are only usable within Dungeons should probably just work using in-client RNG (although it would still be logged).

**Comment by NoiSek** [comment]
Another alternative to the above is just to wait until the end of the dungeon to generate those kinds of rewards. I've always enjoyed the end-fight / end-dungeon screens that pop up, blast your score, give you a rating, then tell you what loot you got.

I also really like the idea of beating a dungeon first, and then being presented at the end of the level with a bunch of chests that contain loot from the monsters you have deaded.

Could even do some super fancy exploding stars, fireworks, and cascading dubloons animations to give users that extra reward feeling.

The notion of having to actually stop game flow and pick up items has always been really unattractive to me as a player, so all of the above sounds pretty appealing.

**Comment by tiffz** [comment]
Thanks for explaining! 

I am really for the idea of eliminating monster drops. I feel like they tend to be kind of artificial and not really make much sense anyways. Perhaps, the items/rewards at the end could be flavored as something like locals come and thank you for clearing the area of monsters if we wanted.

**Pull request merged**